### PR TITLE
[pallas] align interpreter load/store with masked behaviour

### LINKED ISSUE #21143: [pallas] Interpreter mismatch for masked OOB indexing
### Description

For triton (if I have read this correctly) masked load/stores do not occur. So you can request to load/store to an index OOB for ref if that is masked. The current interpreter uses dynamic_slices/dynamic_slice_updates where masked updates are applied. In line with the 'always be in bounds' design in JAX if you index a slice that overruns the edge of the array it will be shifted to be valid (if possible). This leads to a disconnect in interpreter and Pallas outputs.

I know Triton is not Pallas, have you changed the desired behaviour for these cases in Pallas? - in which case this isn't a bug but needs documenting.

I've added a pull request fixing this with some tests https://github.com/google/jax/pull/21298

Here is a [colab](https://colab.research.google.com/drive/1zKJGob-xrKa3Zfbe22m6cSY8V1e9cNgK?usp=sharing) minimal reproduction with shifts in load indices.
```python
from functools import partial
import jax
from jax import numpy as jnp, jit
from jax.experimental import pallas as pl

def masked_load_pallas_kernel(x_ref, o_ref):
  i = jnp.array(3)
  mask = jnp.arange(x_ref.shape[0]) + i < x_ref.shape[0]
  x = pl.load(x_ref, pl.dslice(i, mask.shape[0]), mask=mask, other=-1)
  o_ref[:] = x

@partial(jit, static_argnames=('interpret',))
def masked_load(x: jax.Array, interpret: bool=True):
  return pl.pallas_call(masked_load_pallas_kernel,
                        out_shape=jax.ShapeDtypeStruct(x.shape, x.dtype),
                        interpret = interpret,
                        )(x)

x = jnp.arange(16)
print(f'Input:\nx:\n{x}\n\nOutput:')
for interpret in (True, False):
  print(f'Interpret: {interpret}\n{masked_load(x, interpret=interpret)}')
```

```
Input:
x:
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]

Output:
Interpret: True
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 -1 -1 -1]
Interpret: False
[ 3  4  5  6  7  8  9 10 11 12 13 14 15 -1 -1 -1]
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.26
jaxlib: 0.4.26
numpy:  1.25.2
python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='b70fe499e42d', release='6.1.58+', version='#1 SMP PREEMPT_DYNAMIC Sat Nov 18 15:31:17 UTC 2023', machine='x86_64')


$ nvidia-smi
Thu May  9 08:55:06 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA L4                      Off | 00000000:00:03.0 Off |                    0 |
| N/A   63C    P0              30W /  72W |  17235MiB / 23034MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
+---------------------------------------------------------------------------------------+
```

(Problem persists in 0.4.28)

### PR DESCRIPTION
**Pull request by oliverdutton**
(and adds stride support)
 
 Fixes https://github.com/google/jax/issues/21143

Implements a jittable masked gather/scatter where for load/store/swap any masked indexing does not occur.

For load it sets any masked indexing to index to the first element in the array instead.   
For swap(/store) it also sets masked indexing to the first element (and then deals with special rules to make sure the first element is dealt with correctly)   

The currently used dynamic_slices are replaced with explicit index materialisation and gathers/scatters.
The advantage of doing it this way is that you can combine it with checkify(f, errors=checkify.index_checks) in interpreter mode to check for any unmasked OOB indexing which is (I think, and believe should be) undefined behaviour.

[apologies this is a reopening of a previous request I'd done badly having not checked contributing.md]

### CODE REVIEW

**Comment by oliverdutton** [comment]
I think this is a more general solution to https://github.com/google/jax/pull/21180 @justinjfu 

This covers more incorrect behaviour than padding and allows checkify to be used for indexing checks (rather than covering a subsection of possible user mistakes by padding and identifying by NaN checks)

File: jax/_src/pallas/primitives.py [hunk 1]
```diff
@@ -407,34 +413,42 @@ def _swap_discharge_rule(in_avals, out_avals, *args_flat, args_tree, **_):
...
+      slice_starts = [s.start if isinstance(s, Slice) else s for s in indices]
+      slice_sizes = tuple(s.size if isinstance(s, Slice) else 1 for s in indices)
+      slice_strides = tuple(s.stride if isinstance(s, Slice) else 1 for s in indices)
+      indices = [start + lax.iota(jnp.int32, size) * stride for (start, size, stride)
+                              in zip(slice_starts, slice_sizes, slice_strides)]
+      indices = tuple(jnp.meshgrid(*indices, indexing='ij'))
+
+    if (mask is not None):
+      # masked indexs do not occur, we simulate by using first element in array as dummy
+      is_first_element = (jnp.stack(indices) == 0).all(0)
```
Comment (justinjfu) [inline jax/_src/pallas/primitives.py:436]: Same question here: why do we need to add special case handling to swap using the first element? Could we not solely rely on a jnp.where(mask, ...) to control which values get swapped as done in the original code?

**Review by justinjfu — COMMENTED** [review]
Thanks for the PR! Really appreciate the fixes.

File: jax/_src/pallas/primitives.py [hunk 2]
```diff
@@ -303,19 +303,25 @@ def _load_discharge_rule(in_avals, out_avals, *args_flat, args_tree, **_):
...
-    for s in idx.indices:
-      if isinstance(s, Slice) and s.stride > 1:
-        raise NotImplementedError("Unimplemented stride support.")
     indices = idx.indices
     scalar_dims = [not isinstance(s, Slice) and not s.shape for s in indices]
     slice_starts = [s.start if isinstance(s, Slice) else s for s in indices]
     slice_sizes = tuple(s.size if isinstance(s, Slice) else 1 for s in indices)
-    out_ones = lax.dynamic_slice(ref, slice_starts, slice_sizes=slice_sizes)
+    slice_strides = tuple(s.stride if isinstance(s, Slice) else 1 for s in indices)
+    indices = [start + lax.iota(jnp.int32, size) * stride for (start, size, stride)
```
[the remark this answers is not in the document]
Comment (oliverdutton) [reply jax/_src/pallas/primitives.py:311]: Agreed, it could be slow. The key advantage is that interpreter code (here) on host will produce the same behaviour as the compiled device code.

**Dynamic slice (or dynamic_update_slice) leads to incorrect behaviour** due to XLAs shifting rules. Noted as:
> the potentially surprising behavior for the case where the requested slice overruns the bounds of the array; in this case the start index is adjusted to return a slice of the requested size:
(https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dynamic_slice.html)

an example is given in https://github.com/google/jax/issues/21143

The use of a gather instead of the dynamic slice fixes that incorrect behaviour

File: jax/_src/pallas/primitives.py [hunk 3]
```diff
@@ -303,19 +303,25 @@ def _load_discharge_rule(in_avals, out_avals, *args_flat, args_tree, **_):
...
+                            in zip(slice_starts, slice_sizes, slice_strides)]
+    indices = tuple(jnp.meshgrid(*indices, indexing='ij'))
+    if mask is not None:
+      # masked loads set to index first element in array
+      indices = tuple(jnp.where(mask, indexs, 0) for indexs in indices)
```
[the remark this answers is not in the document]
Comment (oliverdutton) [reply jax/_src/pallas/primitives.py:316]: It's true these values will be masked away on 328. The masked indices being set to the first element is to match the described behaviour of masked loads more closely, the documentation says
> Triton avoids the read from/write to memory if it’s masked

as part of   
> Masking is important when doing out-of-bounds loads/stores. The operational semantics of masking can be compiler-determined (if we understand the documentation properly, Triton avoids the read from/write to memory if it’s masked).
(https://jax.readthedocs.io/en/latest/pallas/design.html)

Those masked indices might be OOB values and should not be indexed at. By diverting them to the first index this is the best jittable approximation I could make. 

The link to checkify is that this change makes it possible to distinguish between an unmasked OOB indexing that will actually happen on device and a a masked OOB indexing that will.
e.g. for Ref of shape (4,) and indices of [1,2,3,7], mask of [True,True,True,False] is a completely valid load, however running checkify check_nan's on the interpretted code would throw an OOB error without this fix as the indexing at 7 would 'physically' occur, the divert to 0 avoids this.

File: jax/_src/pallas/primitives.py (hunk 1 above)
Comment (oliverdutton) [reply jax/_src/pallas/primitives.py:436]: To mirror device behaviour, Ref is not accessed at masked indices. The best jittable approximation I found was to divert masked indices to index at the first element, this leads to masked indexing not occurring   which is often OOB indexing, while the first element should never be OOB.

File: tests/pallas/pallas_test.py [hunk 4]
```diff
@@ -493,6 +493,85 @@ def masked_swap(_, _2, mask_ref, x_ref, y_ref):
...
+        input_output_aliases={0: 0, 1: 1},
+    )
+    def masked_swap(_, _2, mask_ref, x_ref, y_ref):
+      x = x_ref[:]
+      y = pl.swap(y_ref, (slice(None),), x, mask=mask_ref[:])
+      x_ref[:] = y
+
+    x = random.normal(random.key(0), (m, n))
+    y = random.normal(random.key(1), (m, n))
+    mask = jnp.zeros((m, n), dtype=bool)
```
[the remark this answers is not in the document]
Comment (oliverdutton) [reply tests/pallas/pallas_test.py:512]: Yep, it's a special case of the mask which could cause problems.

Will merge together

File: tests/pallas/pallas_test.py [hunk 5]
```diff
@@ -493,6 +493,85 @@ def masked_swap(_, _2, mask_ref, x_ref, y_ref):
...
+    mask = indices < m
+    out = masked_oob_swap_slice(x, y, mask, slice_start)
+
+    # the unjittable masked indexing equivalent
+    unmasked_idx = indices[mask]
+    x_new = x.at[mask].set(y[unmasked_idx])
+    y_new = y.at[unmasked_idx].set(x[mask])
+    np.testing.assert_array_equal(out[0], x_new)
+    np.testing.assert_array_equal(out[1], y_new)
+
```
[the remark this answers is not in the document]
Comment (oliverdutton) [reply tests/pallas/pallas_test.py:574]: Will do

File: jax/_src/pallas/primitives.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinjfu) [reply jax/_src/pallas/primitives.py:316]: The cleaner way to handle this for checkify is probably to add a custom checkify rule for the load and swap primitives (for example, see https://github.com/google/jax/blob/main/jax/_src/checkify.py#L608). The interpret mode discharge rule is primarily responsible for returning the correct values.

What I propose is this change:
1) In the discharge rule for load and swap, simply pad the inputs to the maximum shape with jax.numpy.pad at the beginning of the call, then you can leave the rest of the logic unchanged.
2) Move the explicit indexing logic to a custom checkify rule for load & swap which will check if there are any unmasked indices that are OOB. If they exist, return the OOB error. I believe checkify only checks for NaNs for a subset of arithmetic ops, so you wouldn't need to do NaN checking here.

I think this would keep the discharge implementation simple and fast but also be compatible with checkify & jit.

[the remark this answers is not in the document]
Comment (oliverdutton) [reply jax/_src/pallas/primitives.py:316]: Correctness 
1. In the discharge rule for load and swap, simply pad the inputs to the maximum shape with jax.numpy.pad at the beginning of the call, then you can leave the rest of the logic unchanged.
This equates to padding by the slice size in each dimension. Agreed this will give correct output without performance hit. Will implement.

Debugging interaction with masked OOBs
2. Move the explicit indexing logic to a custom checkify rule for load & swap which will check if there are any unmasked indices that are OOB. If they exist, return the OOB error. I believe checkify only checks for NaNs for a subset of arithmetic ops, so you wouldn't need to do NaN checking here.
I think it makes sense for that to be dealt with by the NaN/OOB checkers. Will look into how to do that and push that into a separate PR to this one. 

This PR will be just to fix #21143 and do 1.

File: tests/pallas/pallas_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (oliverdutton) [reply tests/pallas/pallas_test.py:545]: Added as `test_masked_oob_load_store_slice`

File: jax/_src/pallas/primitives.py [hunk 6]
```diff
@@ -294,6 +294,43 @@ def _load_jvp(primals, tangents, args_tree, **params):
...
+  padding_value = _uninitialized_value(shape=(), dtype=value.dtype)
+  value = lax.pad(value,
+                  padding_config=padding_config,
+                  padding_value=padding_value)
+  return value
+
+_unpad_values_to_avoid_dynamic_slice_oob_shift = partial(
+  _pad_values_to_avoid_dynamic_slice_oob_shift, unpad=True)
+
+def _uninitialized_value(shape, dtype):
```
Comment (oliverdutton) [inline jax/_src/pallas/primitives.py:328]: This is a duplicate of https://github.com/google/jax/blob/1e48adc6982c2c774daf349d6d6574a1d041d772/jax/_src/pallas/pallas_call.py#L112-L117
is this okay, what do you suggest?

**Comment by oliverdutton** [comment]
This PR is now solely aimed at fixing dynamic slice store/load operations in the interpreter identified in https://github.com/google/jax/issues/21143. In the interpreter it pads the arrays with uninitialised values so dynamic slices are never pushed back into the 'true' array.

Tests for both load and store are added. I've confirmed they error pre this commit and are fixed by this change (as is the original bug report).

File: jax/_src/pallas/primitives.py (hunk 6 above)
Comment (justinjfu) [reply jax/_src/pallas/primitives.py:328]: I think it's fine to make this a public function here and import this function from pallas_call.

File: tests/pallas/pallas_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (justinjfu) [reply tests/pallas/pallas_test.py:587]: Thanks!

File: jax/_src/pallas/primitives.py [hunk 7]
```diff
@@ -298,6 +298,39 @@ def _load_jvp(primals, tangents, args_tree, **params):
...
+
+  For example, if arr is [1.,2.,3.,4.] and a slice of size 4, start index 2 is
+  requested then the result will be [3.,4.,NaN,NaN] after padding, rather than
+  [1.,2.,3.,4.] from the unpadded array
+
+  unpad=True performs the inverse operation
+  """
+
+  padding_config = tuple((0, slice_size, 0) for slice_size in slice_sizes)
+  if (unpad):
```
Comment (justinjfu) [inline jax/_src/pallas/primitives.py:316]: nit: Remove parentheses

**Review by justinjfu — CHANGES_REQUESTED** [review]
Thanks! Looks great, just some minor comments.

File: jax/_src/pallas/primitives.py (hunk 6 above)
Comment (oliverdutton) [reply jax/_src/pallas/primitives.py]: Done (have also added behaviour for bool, where it counts False as uninitialised.)

File: jax/_src/pallas/primitives.py (hunk 1 above)
Comment (oliverdutton) [reply jax/_src/pallas/primitives.py]: Dealt with in other comments

**Comment by oliverdutton** [comment]
Comments implemented and now should be ready

**Review by justinjfu — APPROVED** [review]
Approved.
